charging_station.py

```diff
@@ -0,0 +1,156 @@
+        self.load_battery_periode = {"fast" : np.zeros((self.nb_t,2)),"slow" : np.zeros((self.nb_t,2))} # How the player wants to charge/discharge the veicules
+        self.battery_stock = {"slow" : np.zeros((self.nb_t+1,2)), "fast" : np.zeros((self.nb_t+1,2))} # State of the batteries
+        self.nb_fast_max = 2 # Number of Stations Fasts and Lows max
+        self.nb_slow_max = 2
+        self.nb_slow = 2 # Number of Stations Fast and Slow currently used
+        self.nb_fast = 2
+        self.pmax_fast = 22
+        self.pmax_slow = 3
+        self.cmax = 40*4 # Maximal capacity of the CS when the 4 slots are used
+        self.depart = {"slow" : np.zeros(2), "fast" : np.zeros(2)} # Time of departure of every cars
+        self.arrival = {"slow" : np.zeros(2), "fast" : np.zeros(2)} # Time of arrival of every cars
+        self.here = {"slow" : np.ones(2), "fast" : np.ones(2)}
+        self.imbalance=[]
+
+    def update_battery_stock(self,time,load_battery):
+        self.nb_cars(time) # We check what cars is here
+        p_max = {"slow" : [3*self.here["slow"][0],3*self.here["slow"][1]], "fast" : [22*self.here["fast"][0],22*self.here["fast"][1]]}
+        c_max = {"slow" : [40*self.here["slow"][0],40*self.here["slow"][1]], "fast" : [40*self.here["fast"][0],40*self.here["fast"][1]]}
+        # p_max and c_max depend on whether the car is here or not.
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if abs(load_battery[speed][i]) >= p_max[speed][i]:
+                    load_battery[speed][i] = p_max[speed][i]*np.sign(load_battery[speed][i])
+            # Can't put more power than p_max
+            new_stock = { "slow" : [0,0], "fast" : [0,0] }
+
+            new_stock["slow"][0] = self.battery_stock["slow"][time][0] + (self.efficiency*max(0,load_battery["slow"][0])+min(0,load_battery["slow"][0])/self.efficiency)*self.dt
+            new_stock["slow"][1] = self.battery_stock["slow"][time][1] + (self.efficiency*max(0,load_battery["slow"][1])+min(0,load_battery["slow"][1])/self.efficiency)*self.dt
+            # We update the new stock of each batteries "slow"
+
+            new_stock["fast"][0]=self.battery_stock["fast"][time][0] + (self.efficiency*max(0,load_battery["fast"][0])+min(0,load_battery["fast"][0])/self.efficiency)*self.dt
+            new_stock["fast"][1]=self.battery_stock["fast"][time][1] + (self.efficiency*max(0,load_battery["fast"][1])+min(0,load_battery["fast"][1])/self.efficiency)*self.dt
+            # We update the new stock of each batteries "fast"
+
+
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if new_stock[speed][i] < 0:
+                    load_battery[speed][i] = -(self.battery_stock[speed][time][i])/(self.efficiency*self.dt)
+                    new_stock[speed][i] = 0
+            # We can't discharge the batterie under 0
+
+                elif new_stock[speed][i] > c_max[speed][i]:
```

Comment:
This condition will generate false negative charging loads when EV are not present (c_max=0). The case of EV not present should be treated carefully in this function.

Reply:
oui c'est pour ça qu'on a observé un -256 dans le load au moment ou les voitures partaient. C'est pourquoi j'ai due remettre la partie de code : 
for speed in ["slow","fast"] :
            for i in range(2):
                if abs(load_battery[speed][i]) >= p_max[speed][i] :
                    load_battery[speed][i] = p_max[speed][i]*np.sign(load_battery[speed][i])
L64 
qui permet de faire une seconde vérification et donc normalement règle le problème, certes c'est un peu bricolé mais normalement ça fonctionne

Reply:
Selon moi, pour que le code soit robuste il faudrait s'assurer de deux points : 1. que les décisions de charge prises fixent une "load" à 0 quand les VE ne sont pas là ; 2. que la mise à jour du stock batterie soit cohérente même si une "load" non nulle est fournie alors qu'un VE n'est pas là. Sur ce second point je ne pense pas que cela soit le cas dans le code actuel qui peut générer des load_battery non nulles (L61) même si la seconde vérification (L66) permet que celles-ci respectent les contraintes de pmax. Je suggèrerais donc de traiter à part le cas où les VEs ne sont pas là en gardant le même état de stock batterie et en fixant la load à 0